lib/SimpleTest/SimpleTestFinder.cpp

```diff
@@ -44,42 +61,137 @@ std::vector<std::unique_ptr<Test>> SimpleTestFinder::findTests(Context &Ctx) {
-        for (auto &Fn : M->getModule()->getFunctionList()) {
-          if (Fn.getName() == F->getName()) {
+    for (inst_iterator I = inst_begin(traverseeFunction),
+         E = inst_end(traverseeFunction); I != E; ++I) {
 
-            // Ignore export declarations.
-            if (!Fn.empty()) {
-              Logger::info()
-                  << "SimpleTestFinder::findTestees - found function "
-                  << Fn.getName() << '\n';
+      auto instruction = &*I;
 
-              testees.push_back(std::make_pair(&Fn, 0));
-            }
+      CallInst *callInstruction = dyn_cast<CallInst>(instruction);
+      if (callInstruction == nullptr) {
+        continue;
+      }
+
+      int callOperandIndex = callInstruction->getNumOperands() - 1;
+      Value *callOperand = callInstruction->getOperand(callOperandIndex);
+      Function *functionOperand = dyn_cast<Function>(callOperand);
+
+      if (!functionOperand) {
+        continue;
+      }
+
+      /// Two modules may have static function with the same name, e.g.:
+      ///
+      ///   // ModuleA
+      ///   define range() {
+      ///     // ...
+      ///   }
+      ///
+      ///   define test_A() {
+      ///     call range()
+      ///   }
+      ///
+      ///   // ModuleB
+      ///   define range() {
+      ///     // ...
+      ///   }
+      ///
+      ///   define test_B() {
+      ///     call range()
+      ///   }
+      ///
+      /// Depending on the order of processing either `range` from `A` or `B`
+      /// will be added to the `context`, hence we may find function `range`
+      /// from module `B` while processing body of the `test_A`.
+      /// To avoid this problem we first look for function inside of a current
+      /// module.
+      ///
+      /// FIXME: Context should report if a function being added already exist
+      /// FIXME: What other problems such behaviour may bring?
+
+      Function *definedFunction = testBodyModule->getFunction(functionOperand->getName());
+
+      if (!definedFunction || definedFunction->isDeclaration()) {
+        definedFunction = Ctx.lookupDefinedFunction(functionOperand->getName());
+      }
+
+      if (definedFunction) {
+        auto functionWasNotProcessed = checkedFunctions.find(definedFunction) == checkedFunctions.end();
+        checkedFunctions.insert(definedFunction);
+
+        if (functionWasNotProcessed) {
+          /// Filtering
+          if (shouldSkipDefinedFunction(definedFunction)) {
+            continue;
           }
+
+          /// The code below is not actually correct
+          /// For each C++ constructor compiler can generate up to three
+          /// functions*. Which means that the distance might be incorrect
+          /// We need to find a clever way to fix this problem
+          ///
+          /// * Here is a good overview of what's going on:
+          /// http://stackoverflow.com/a/6921467/829116
+          ///
+          traversees.push(new Testee(definedFunction,
```

Comment:
I will address all of the comments tomorrow. For now one question about this: later this allocated testee gets pushed to `std::vector<std::unique_ptr<Testee>> testees;` via

```
std::unique_ptr<Testee> uniqueTestee(traversee);
testees.push_back(std::move(uniqueTestee));
```

doesn't that mean that this allocation will be managed by vector of unique pointers?

That's what my assumption is about how unique pointers work. That's not the only place in the code where I and Alexey exploit this behaviour. Are we wrong?